File: ios/player/PlaybackStateEmissionGate.swift
import Foundation

struct PlaybackStateFingerprint: Equatable, Sendable {
  let status: Int
  let currentTimeBits: UInt64
  let durationBits: UInt64
  let bufferDurationBits: UInt64
  let bufferedPositionBits: UInt64
  let rateBits: UInt64
  let isPlaying: Bool
  let isBuffering: Bool
  let isVisualReady: Bool
  let errorCode: Int?
  let errorMessageHash: UInt64?

  init(
    status: Int,
    currentTime: Double,
    duration: Double,
    bufferDuration: Double,
    bufferedPosition: Double,
    rate: Double,
    isPlaying: Bool,
    isBuffering: Bool,
    isVisualReady: Bool,
    errorCode: Int?,
    errorMessageHash: UInt64?
  ) {
    self.status = status
    self.currentTimeBits = Self.normalizedBits(currentTime)
    self.durationBits = Self.normalizedBits(duration)
    self.bufferDurationBits = Self.normalizedBits(bufferDuration)
    self.bufferedPositionBits = Self.normalizedBits(bufferedPosition)
    self.rateBits = Self.normalizedBits(rate)
    self.isPlaying = isPlaying
    self.isBuffering = isBuffering
    self.isVisualReady = isVisualReady
    self.errorCode = errorCode
    self.errorMessageHash = errorMessageHash
  }

  private static func normalizedBits(_ value: Double) -> UInt64 {
    if value.isNaN {
      return UInt64.max
    }
    return value.bitPattern
  }
}

final class PlaybackStateEmissionGate {
  private var lastFingerprint: PlaybackStateFingerprint?
  private let lock = NSLock()

  func shouldEmit(fingerprint: PlaybackStateFingerprint) -> Bool {
    lock.lock()
    defer { lock.unlock() }
    if fingerprint == lastFingerprint {
      return false
    }
    lastFingerprint = fingerprint
    return true
  }

  func reset() {
    lock.lock()
    lastFingerprint = nil
    lock.unlock()
  }
}

File: android/src/main/java/com/nitroplay/player/PlaybackStateEmissionGate.kt
package com.margelo.nitro.video

internal data class PlaybackStateFingerprint(
  val status: NitroPlayerStatus,
  val currentTimeBits: Long,
  val durationBits: Long,
  val bufferDurationBits: Long,
  val bufferedPositionBits: Long,
  val rateBits: Long,
  val isPlaying: Boolean,
  val isBuffering: Boolean,
  val isVisualReady: Boolean,
  val errorCode: NitroPlayerErrorCode?,
  val errorMessageHash: Int?
) {
  companion object {
    fun from(state: PlaybackState): PlaybackStateFingerprint {
      val error = state.error?.asSecondOrNull()
      return PlaybackStateFingerprint(
        status = state.status,
        currentTimeBits = state.currentTime.toBits(),
        durationBits = state.duration.toBits(),
        bufferDurationBits = state.bufferDuration.toBits(),
        bufferedPositionBits = state.bufferedPosition.toBits(),
        rateBits = state.rate.toBits(),
        isPlaying = state.isPlaying,
        isBuffering = state.isBuffering,
        isVisualReady = state.isVisualReady,
        errorCode = error?.code,
        errorMessageHash = error?.message?.hashCode()
      )
    }
  }
}

internal class PlaybackStateEmissionGate {
  private val lock = Any()
  private var lastFingerprint: PlaybackStateFingerprint? = null

  fun shouldEmit(fingerprint: PlaybackStateFingerprint): Boolean {
    synchronized(lock) {
      if (fingerprint == lastFingerprint) {
        return false
      }
      lastFingerprint = fingerprint
      return true
    }
  }

  fun reset() {
    synchronized(lock) {
      lastFingerprint = null
    }
  }
}
